### PR TITLE
Products controller

### PR DESCRIPTION
**Pull request by christianpaez**
## Description

Adds controller and routing helper for Product resource

## Tickets

https://flatirons.atlassian.net/browse/FS-17

### CODE REVIEW

File: lib/flatirons/saas/rails/routes.rb [hunk 1]
```diff
@@ -92,30 +112,54 @@ def extract_resource_ref(resource, options) # :nodoc:
...
-
       ensure_devise! resource, symbol, klass
-      ensure_subscriptable! klass
+      ensure_modulable! klass, 'subscriptable'
 
       Flatirons::Saas::Rails::Subscription.mappings[symbol] = { symbol: symbol, name: name, path: path, klass: klass, resource: resource }
       Flatirons::Saas::Rails::Subscription.mappings[symbol]
     end
 
+    def extract_products_resource_ref(resource, options) # :nodoc:
```
Comment (christianpaez) [inline lib/flatirons/saas/rails/routes.rb]: Todo: Refactor dup code

File: app/controllers/flatirons/saas/products_controller.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      opts[:scope] = symbol
+      @current_resource = warden.authenticate!(opts)
+      raise_current_resource_mapping_error unless @current_resource.is_a? @mapping[:klass]
+      @current_resource
+    end
+
+    def raise_current_resource_mapping_error
+      raise "Authenticated resource type [#{@current_resource.class}] is not the same of mapped [#{@mapping[:klass]}],"\
+      ' check your subscription_for/devise_for route configuration.'
+    end
```
Comment (gleandroj) [inline app/controllers/flatirons/saas/products_controller.rb]: Can we reuse this logic between subscriptions and products controllers:

File: spec/support/products_for_users.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+shared_context 'products_for_users' do
+  before(:each) do
+    Rails.application.routes.draw do
+      devise_for :dummy_users
+      devise_scope :dummy_user do
+        products_for :dummy_users, resource_class_name: 'Flatirons::Saas::Product'
+      end
+    end
+  end
+end
```
Comment (gleandroj) [inline spec/support/products_for_users.rb:12]: Maybe you can put this on routes.rb of dummy project like I did to subscriptions

File: spec/support/subscription_for_users.rb [hunk 4]
```diff
@@ -3,9 +3,9 @@
...
     Rails.application.routes.draw do
-      devise_for :users
-      devise_scope :user do
-        subscription_for :users
+      devise_for :dummy_users
+      devise_scope :dummy_user do
+        subscription_for :dummy_users
       end
     end
   end
```
Comment (gleandroj) [inline spec/support/subscription_for_users.rb:11]: It can be deleted

**Review by gleandroj — APPROVED** [review]

File: spec/support/subscription_for_users.rb (hunk 4 above)
Comment (christianpaez) [reply spec/support/subscription_for_users.rb:11]: we need it for testing, if we delete routes from /spec/dummy subscription tests will fail

**Pull request merged**